pkg/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/fission/fission/pkg/fission-cli/util"
+)
+
+type CategoryID string
+
+const (
+	Kubernetes      CategoryID = "kubernetes"
+	FissionServices CategoryID = "fission-services"
+	FissionVersion  CategoryID = "fission-version"
+)
+
+var (
+	okStatus   = color.New(color.FgGreen, color.Bold).SprintFunc()("\u221A") // √
+	failStatus = color.New(color.FgRed, color.Bold).SprintFunc()("\u00D7")   // ×
+)
+
+// Category is a group of checkers
+type Category struct {
+	ID       CategoryID
+	checkers []Checker
+	enabled  bool
+}
+
+type Checker struct {
+	successMsg string
+	check      func() error
+}
+
+type Options struct {
+	KubeContext   string
+	FissionClient client.Interface
+}
+
+type HealthChecker struct {
+	categories []*Category
+	*Options
+
+	kubeAPI          *kubernetes.Clientset
+	fissionNamespace string
+}
+
+func isCompatibleVersion(minimalRequirementVersion [3]int, actualVersion [3]int) bool {
+	if minimalRequirementVersion[0] < actualVersion[0] {
+		return true
+	}
+
+	if (minimalRequirementVersion[0] == actualVersion[0]) && minimalRequirementVersion[1] < actualVersion[1] {
+		return true
+	}
+
+	if (minimalRequirementVersion[0] == actualVersion[0]) && (minimalRequirementVersion[1] == actualVersion[1]) && (minimalRequirementVersion[2] <= actualVersion[2]) {
+		return true
+	}
+
+	return false
+}
+
+func (hc *HealthChecker) CheckKubeVersion() (err error) {
+
+	version, err := hc.kubeAPI.ServerVersion()
+	if err != nil {
+		return err
+	}
+
+	major, _ := strconv.Atoi(version.Major)
+	minor, _ := strconv.Atoi(version.Minor)
+	apiVersion := [3]int{major, minor, 0}
+	minAPIVersion := [3]int{1, 19, 0}
+
+	if !isCompatibleVersion(minAPIVersion, apiVersion) {
+		return fmt.Errorf("kubernetes is on version %d.%d.%d, but version %d.%d.%d or more recent is required",
+			apiVersion[0], apiVersion[1], apiVersion[2],
+			minAPIVersion[0], minAPIVersion[1], minAPIVersion[2])
+	}
+
+	return nil
+}
+
+func (hc *HealthChecker) CheckServiceStatus(namespace string, name string) (err error) {
+	depl, err := hc.kubeAPI.AppsV1().Deployments(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get %s deployment status", name)
+	}
+
+	if depl.Status.UnavailableReplicas > 0 || depl.Status.Replicas == 0 {
+		return fmt.Errorf("%s deployment is not running", name)
+	}
+
+	_, err = hc.kubeAPI.CoreV1().Services(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get %s service status", name)
+	}
+
+	return nil
+}
+
+func (hc *HealthChecker) CheckFissionVersion() error {
+	ver := util.GetVersion(hc.FissionClient)
+
+	major, _ := strconv.Atoi(strings.Split(ver.Client["fission/core"].Version, ".")[0])
+	minor, _ := strconv.Atoi(strings.Split(ver.Client["fission/core"].Version, ".")[1])
+	fix, _ := strconv.Atoi(strings.Split(ver.Client["fission/core"].Version, ".")[2])
+
+	apiVersion := [3]int{major, minor, fix}
+	minAPIVersion := [3]int{1, 15, 1}
```

Comment:
Latest fission released version is hardcoded. I want to confirm whether this should go inside `pkg/apis/core/v1/const.go`

Reply:
We should check client vs server version here. Ideally we should require hardcoding.
client version is embedded in build info, where as server info is fetched from fission.
Client version should match with server version.

Reply:
Please refere server and client versions from https://github.com/fission/fission/blob/3ac188124e348c2463a9e8b46bae9ab8967ce255/pkg/fission-cli/util/util.go#L172